File: Online judge/Boj/Python/16198.py
ans = 0
def dfs(gu, sum):
  global ans
  if len(gu) == 2:
    if ans < sum:
      ans = sum
    return
  for x in range(1, len(gu)-1):
    num = gu[x]
    sum += gu[x-1]*gu[x+1]
    del gu[x]
    dfs(gu, sum)
    sum -= gu[x-1]*gu[x+1]
    gu.insert(x, num)

N = input()
lis = list(map(int, input().split()))
dfs(lis, 0)
print(ans)

File: Online judge/Boj/C++/16198.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define endl '\n'

using namespace std;

int N, ans;

void dfs(vector<int> energy, int sum) {
	if (energy.size() == 2) {
		ans = max(ans, sum);
		return;
	}
	for (int i = 1; i < energy.size() - 1; i++) {
		int num = energy[i];
		sum += energy[i - 1] * energy[i + 1];
		energy.erase(energy.begin() + i);
		dfs(energy, sum);
		energy.insert(energy.begin() + i, num);
		sum -= energy[i - 1] * energy[i + 1];
	}
}

int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);
	
	vector<int> v;
	cin >> N;
	for (int i = 0; i < N; i++) {
		int n;
		cin >> n;
		v.push_back(n);
	}

	dfs(v, 0);
	cout << ans << endl;
}